notmuch/query.c

```diff
@@ -32,6 +32,56 @@
 #include "mutt/lib.h"
 #include "query.h"
 
+/**
+ * nm_parse_type_from_query - Parse a query type out of a query
+ * @param buf   Buffer for URL
+ * @retval Notmuch query type.  Default: #NM_QUERY_TYPE_MESGS
+ *
+ * If a user writes a query for a vfolder and includes a type= statement, that
+ * type= will be encoded, which Notmuch will treat as part of the query=
+ * statement. This method will remove the type= and return its corresponding
+ * NmQueryType representation.
+ */
+enum NmQueryType nm_parse_type_from_query(char *buf)
+{
+  if (!buf)
+    return NM_QUERY_TYPE_MESGS;
+
+  // The six variations of how type= could appear.
+  const char *variants[6] = { "&type=threads", "&type=messages",
+                              "type=threads&", "type=messages&",
+                              "type=threads",  "type=messages" };
+
+  enum NmQueryType query_type = NM_QUERY_TYPE_MESGS;
+  int variants_size = mutt_array_size(variants);
+  for (int i = 0; i < variants_size; i++)
+  {
+    if (mutt_istr_find(buf, variants[i]) != NULL)
+    {
+      // variants[] is setup such that type can be determined via modulo 2.
+      query_type = ((i % 2) == 0) ? NM_QUERY_TYPE_THREADS : NM_QUERY_TYPE_MESGS;
+
+      mutt_istr_remall(buf, variants[i]);
+    }
+  }
+
+  return query_type;
+}
+
+/**
+ * query_type_to_string - Turn a query type into a string
+ * @param query_type Query type
+ * @retval ptr String
+ *
+ * @note This is a static string and must not be freed.
+ */
+const char *nm_query_type_to_string(enum NmQueryType query_type)
+{
+  if (query_type == NM_QUERY_TYPE_THREADS)
+    return "threads";
+  return "messages";
+}
+
 /**
  * nm_string_to_query_type - Lookup a query type
  *
```

Comment:
I've had another think about this one...
I'm not happy with the error message being in it.

I think the function should return an error value if the lookup fails,
and allow the caller to decide whether an error message should be shown.

(this would also mean that we don't need to mess around with `MuttLogger` in the tests)

Reply:
> I'm not happy with the error message being in it.

I'm of the same opinion. Utility functions should

> I think the function should return an error value if the lookup fails,

What's the best way to accomplish that? Extend `NmQueryType` with `NM_QUERY_TYPE_UNKNOWN` and return that for higher-level code to handle it? Or am I missing something obvious?

Reply:
> Extend `NmQueryType` with `NM_QUERY_TYPE_UNKNOWN`

Yes, that's seem best.

> return that for higher-level code to handle it?

Exactly.  A small wrapper function that, on error, logs and chooses the default.